nvidia-persistenced-installer/nvidia_persistenced_installer.go

```
@@ -0,0 +1,167 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"os"
+	"os/exec"
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/golang/glog"
+)
+
+const (
+	minUVMSupportedVersion = 550
+)
+
+var (
+	containerPathPrefix            = flag.String("container-path", "/usr/local/nvidia", "Path on the container that mounts host nvidia install directory")
+	cgpuConfigFile                 = flag.String("cgpu-config", "/etc/nvidia/confidential_node_type.txt", "File with Confidential Node Type used on Node")
+)
+
+func main() {
+	flag.Parse()
+	ctx := context.Background()
+
+	// Only run persistence daemon on confidential GPU nodes.
+	enabled, err := checkConfidentialGPUEnablement(ctx)
+	if err != nil {
+		glog.ExitContextf(ctx, "parseCGPUConfig failed: %v", err)
+	}
+
+	if enabled {
+		// This is necessary to be able to use nvidia smi from the container to set the GPU to a ready state.
+		if err := updateContainerLdCache(); err != nil {
+			glog.ExitContextf(ctx, "updateContainerLdCache failed: %v", err)
+		}
+	
+		if err := enablePersistenceMode(ctx); err != nil {
+			glog.ExitContextf(ctx, "failed to start persistence mode: %v", err)
+		}
+		// Add small delay before setting the ready state for consistency.
+		// If the workload starts too close to when the persistence daemon has started sometimes there can be errors.
+		time.Sleep(1 * time.Second)
+		if err := setGPUReadyState(ctx); err != nil {
+			glog.ExitContextf(ctx, "failed to set gpu to ready state: %v", err)
+		}
+	} else {
+		glog.InfoContext(ctx, "Confidential GPU is NOT enabled, skipping nvidia persistenced enablement.")
+		// Don't exit as this is intended for a side car which would cause it to restart infinitely.
+	}
+
+	// Need to keep the container running so that the nvidia persistence daemon can keep running.
+	for {
+		glog.InfoContext(ctx, "sleeping");
+		time.Sleep(5 * time.Minute)
+	}
```

Several questions about this 
1. Can persistend fail after it's started? If so how can it be monitored from this container? 
2. Let's not put a log "sleeping" every 5 mins as it might lead to log spam

---

The persistence daemon after its done just sleeps for the rest of the time. So I think it is unlikely to fail after it has started.

I'm not sure if we have a good way to monitor it, but if it does fail later, from my experience things end up broken even if you try to restart it. It usually just leads to having to reboot the node as you can't get the encryption keys back. So I don't think it's worth it to try to monitor it and restart it.

---

Agreed on the log spam, I'll remove the sleep log.

---

Will the GPU device plugin container fail or report unhealthy GPUs if persistenced fails? I worry it'll just silently fail without the observability to tell what's wrong, then it'll be hard for users and us to debug.

---

I think it would silently fail. We might be able to monitor it later with the device plugin using the nvidia manager? For example the GPU would get detached and would probably be in an unhealthy state, so that might get picked up by this? https://github.com/GoogleCloudPlatform/container-engine-accelerators/blob/master/cmd/nvidia_gpu/nvidia_gpu.go#L142

---

Per discussion in chat, I'm fine with merging this as it is now but we probably want to come back later for error detection/handling. Thanks!